File: client/src/method.ts
import { EntryHash, EntryHashB64 } from "@holochain/client"
import { ConfigDimension, DimensionEh } from "./dimension"
import { ConfigResourceDef, ResourceDefEh, ResourceEh } from "./resourceDef"

interface CoreMethod {
    name: string,
    program: Program,
    can_compute_live: boolean,
    requires_validation: boolean,
}

export type Method = CoreMethod & {
  target_resource_def_eh: ResourceDefEh,
  input_dimension_ehs: Array<DimensionEh>,
  output_dimension_eh: DimensionEh,
}

export type RawMethod = CoreMethod & {
  target_resource_def_eh: EntryHash,
  input_dimension_ehs: Array<EntryHash>,
  output_dimension_eh: EntryHash,
}

export type ConfigMethod = CoreMethod & {
    target_resource_def: ConfigResourceDef,
    input_dimensions: Array<ConfigDimension>,
    output_dimension: ConfigDimension,
}

export interface RunMethodInput {
  resource_eh: ResourceEh,
  method_eh: MethodEh,
}
export interface DataSet {
    from: EntryHashB64, // :TODO: ResourceEh?
    data_points: {
        [key: string]: Array<EntryHashB64>, // key cannot be of type Uint8Array, specifying as string for now as we are not currently using `DataSet`
    }
}

export type Program = ProgramSum | ProgramAverage

export interface ProgramSum {
    Sum: null,
}

export interface ProgramAverage {
    Average: null,
}

export type MethodEh = EntryHashB64

File: dnas/sensemaker/zomes/integrity/sensemaker_structs/src/method.rs
use std::collections::BTreeMap;

use hdi::prelude::*;

use crate::{applet::ConfigMethod, Dimension};

#[hdk_entry_helper]
#[derive(Clone)]
pub struct Method {
    pub name: String,
    pub input_dimension_ehs: Vec<EntryHash>, // Validation: make sure it is subjective
    pub output_dimension_eh: EntryHash,      // Validation: make sure it is objective
    pub program: Program,                    // making enum for now, in design doc it is `AST`
    pub can_compute_live: bool,
    pub requires_validation: bool, // if true, DataSet must be committed to be retrievable in the validation The Objective Assesment must have the DataSet.
}

// Used for atomic operation in method creation
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct PartialMethod {
    pub name: String,
    pub input_dimension_ehs: Vec<EntryHash>,
    pub output_dimension_eh: Option<EntryHash>, // This is now nullable
    pub program: Program,                   
    pub can_compute_live: bool,
    pub requires_validation: bool
}

impl TryFrom<ConfigMethod> for Method {
    type Error = WasmError;
    fn try_from(value: ConfigMethod) -> Result<Self, Self::Error> {
        let input_dimension_ehs = value
            .input_dimensions
            .into_iter()
            .map(|config_dimension| {
                let converted_dimension: Dimension = Dimension::try_from(config_dimension)?;
                hash_entry(converted_dimension)
            })
            .collect::<ExternResult<Vec<EntryHash>>>()?;

        let converted_output_dimension: Dimension = Dimension::try_from(value.output_dimension)?;
        let output_dimension_eh = hash_entry(converted_output_dimension)?;

        let method = Method {
            name: value.name,
            input_dimension_ehs,
            output_dimension_eh,
            program: value.program,
            can_compute_live: value.can_compute_live,
            requires_validation: value.requires_validation,
        };
        Ok(method)
    }
}

#[hdk_entry_helper]
#[derive(Clone)]
pub struct DataSet {
    pub from: EntryHash,                                  // method
    pub data_points: BTreeMap<EntryHash, Vec<EntryHash>>, //<DimensionEh, Vec<AssessmentEh>>
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Program {
    Sum,
    Average,
}
